llvm/lib/Target/Mips/CheriPureCapABI.cpp

```diff
@@ -263,6 +261,39 @@ class CheriPureCapABI : public ModulePass, public InstVisitor<CheriPureCapABI> {
       if (AI->isArrayAllocation())
         Size = B.CreateMul(Size, AI->getArraySize());
 
+      if (AI->isStaticAlloca() && ForcedAlignment) {
+        // Pad to ensure bounds don't overlap adjacent objects
+        uint64_t AllocaSize =
+            cast<ConstantInt>(Size)->getValue().getLimitedValue();
+        uint64_t AlignedSize = alignTo(AllocaSize, ForcedAlignment);
```

Comment:
I originally though of adding the padding in the prologepilog code, but I guess changing the IR type might  be simpler.

Reply:
The AArch64 has its own pass which, among other things, does this for MTE (to align each alloca to the the tag granule size) in `AArch64StackTagging::alignAndPadAlloca`, which you may notice bears a remarkable resemblance to the code here.

Reply:
Sounds good. I guess we can always try to upstream a patch that factors out that method to avoid duplicating code.